Source/WebCore/Modules/url-pattern/URLPattern.cpp

```diff
@@ -26,29 +26,247 @@
+    String&& password, String&& hostname, String&& port, String&& pathname, String&& search, String&& hash)
+{
+    URLPatternInit result;
+    result.protocol = protocol;
+    result.username = username;
+    result.password = password;
+    result.hostname = hostname;
+    result.port = port;
+    result.pathname = pathname;
+    result.search = search;
+    result.hash = hash;
+
+    URL baseURL;
+
+    if  (!init.baseURL.isNull()) {
+        baseURL = URL(init.baseURL);
+
+        if (!baseURL.isValid() || baseURL.isEmpty())
+            return Exception { ExceptionCode::TypeError, "Invalid baseURL"_s };
+
+        if (init.protocol.isNull())
+            result.protocol = processBaseURLString(baseURL.protocol(), type);
+
+        if (type != BaseURLStringType::Pattern
+            && init.protocol.isNull()
+            && init.hostname.isNull()
+            && init.port.isNull()
+            && init.username.isNull())
+            result.username = processBaseURLString(baseURL.user(), type);
+
+        if (type != BaseURLStringType::Pattern
+            && init.protocol.isNull()
+            && init.hostname.isNull()
+            && init.port.isNull()
+            && init.username.isNull()
+            && init.password.isNull())
+            result.password = processBaseURLString(baseURL.password(), type);
+
+        if (init.protocol.isNull()
+            && init.hostname.isNull()) {
+            result.hostname = processBaseURLString(!baseURL.host().isNull() ? baseURL.host() : StringView { emptyString() }, type);
+        }
+
+        if (init.protocol.isNull()
+            && init.hostname.isNull()
+            && init.port.isNull()) {
+            auto port = baseURL.port();
+            result.port = port ? String::number(*port) : emptyString();
+        }
+
+        if (init.protocol.isNull()
+            && init.hostname.isNull()
+            && init.port.isNull()
+            && init.pathname.isNull()) {
+            result.pathname = processBaseURLString(baseURL.path(), type);
+        }
+
+        if (init.protocol.isNull()
+            && init.hostname.isNull()
+            && init.port.isNull()
+            && init.pathname.isNull()
+            && init.search.isNull()) {
+            result.search = processBaseURLString(baseURL.hasQuery() ? baseURL.query() : StringView { emptyString() }, type);
+        }
+
+        if (init.protocol.isNull()
+            && init.hostname.isNull()
+            && init.port.isNull()
+            && init.pathname.isNull()
+            && init.search.isNull()
+            && init.hash.isNull()) {
+            result.hash = processBaseURLString(baseURL.hasFragmentIdentifier() ? baseURL.fragmentIdentifier() : StringView { emptyString() }, type);
+        }
+    }
+
+    if (!init.protocol.isNull()) {
+        auto protocolResult = canonicalizeProtocol(init.protocol, type);
```

Comment:
It looks like all these canonicalize* functions need `dummyURL` and we are creating it for each function. Can we just create one `dummyURL` and passes its reference to these functions? This will help avoid parsing on URL creation.

Reply:
Chris and Ryosuke discussed it earlier and they said that they wanted to leave as-is for now. 

<img width="836" alt="Screenshot 2024-10-22 at 11 55 20 AM" src="https://github.com/user-attachments/assets/9e06cf35-22a4-452b-818c-914fd52ffee4">